### DISCUSSION TITLE
How do I enter hyper/meh or other key-combos in the GUI?

### DISCUSSION THREAD

**Discussion by ankushg** [nickcoutsos/keymap-editor#50] [Q&A] [answered]
Hey there! I'm trying to make a `hyper` key and a `meh` key. 

I _think_ the way to do this is by using `&mt LC(LS(LALT))` for `meh` and `&mt LC(LS(LA(LGUI)))` for `hyper`

Is this something I can do in the UI?

**Comment by nickcoutsos** [comment, marked as answer]
Hello!

The editor only supports the modifier macros currently described in the ZMK docs so you'd have to do what you've described here. If you're editing your keymap code directly you might even do something like the following

```C
#define MEH LC(LS(LALT))
#define HYPER LC(LS(LA(LGUI)))
```

_however_ the editor may not (it's been a while since I've considered dealing with custom preprocessor macros) parse these. Or if it does, it may replace it when parsing the keymap but won't restore the custom identifier when rewriting the devicetree source. All that to say, I wouldn't currently recommend it. I know that these definitions are used a lot in keymaps, but I see that as the tradeoff between the convenience of writing less code versus the convenience of writing next to no code.

**Reply by ankushg** [reply]
That’s fair!

I wonder if there’s a way for the parser to have some sort of “raw” mode
that it falls back to when it doesn’t recognize what’s in the keymap?

That would let folks do less-official things in the rare cases that they’re
needed

**Reply by nickcoutsos** [reply]
That's a possibility, but there are a lot of tricky questions around it.

It's one thing to have `#define SYM 1` for a layer index and have the editor display but ignore that identifier in the keymap, but I've also seen people use a macro to replace multiple bindings which I can't map to the layout correctly.

**Reply by nickcoutsos** [reply]
We're talking about supporting a "hyper" modifier because that would more convenient than adding up all the individual modifiers, but maybe the other thing to consider is just making it easier to apply modifiers to a key code or to combine a set of modifiers for binding cells that expect a modifier.

My "paradigm" of treating these modifier macros as infinitely nestable function calls was an easy way to structure it but not the best user experience. I've recently made some improvements to the picker components to allow for greater flexibility, so it might be time for me to revisit this anyway.
